zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -217,19 +217,24 @@ mod tests {
         let re = Regex::new(r"\s+").unwrap();
 
         let expected_tables = vec![
-            "CREATE TABLE \"accounts\" (
-                account INTEGER PRIMARY KEY,
-                ufvk TEXT NOT NULL,
+            r#"CREATE TABLE "accounts" (
+                id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                account_type INTEGER NOT NULL DEFAULT 0,
+                hd_seed_fingerprint BLOB,
+                hd_account_index INTEGER,
+                uvk TEXT NOT NULL,
                 birthday_height INTEGER NOT NULL,
-                recover_until_height INTEGER )",
-            "CREATE TABLE addresses (
-                account INTEGER NOT NULL,
+                recover_until_height INTEGER,
+                CHECK ( (account_type = 0 AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL) OR (account_type != 0 AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL) )
```

Comment:
This allows an `account_type` other than 0 or 1; is that intended?

```suggestion
                CHECK ( (account_type = 0 AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL) OR (account_type = 1 AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL) )
```

Reply:
Yes, that was the suggestion I made here: https://github.com/zcash/librustzcash/pull/1175#pullrequestreview-1920990166 (If you disagree with my suggestion, let's talk about it here.)

Reply:
> The CHECK here cannot be changed later without doing another "create new table, copy all the data across, delete old table, rename new table" cycle.

That is pretty much what we always do for migrations. This migration does it five times (for tables `accounts`, `addresses`, `sapling_received_notes`, `sent_notes`, and `utxos`)! It's just harmless boilerplate.

> Given that the only other account types I'm aware of that might be added are non-ZIP-32 types, we should pre-emptively just gate the other half of this check on account_type != 0.

My reasoning was that the CHECK should be as narrow as possible to reflect the *current* database schema. We don't know and shouldn't try to predict what the correct CHECK will be for future types. I see your argument that "the only other account types I'm aware of that might be added are non-ZIP-32 types" and I'm not convinced, because it's quite likely that any new types will need new columns. That would in any case require a modification to the CHECK, to say that those columns are NULL for types 0 and 1, and non-NULL or whatever for the new type.

Not fully checking `account_type`, on the other hand, is not harmless. If we have an `account_type` that is not 0 or 1 in the current schema, what does it mean? It should be treated as database corruption.

Reply:
Btw I agree with your and @nuttycom's reasoning on the magic numbers (https://github.com/zcash/librustzcash/pull/1175#discussion_r1515491709).

Reply:
Ok. I'll switch to using the magic number constants then. As for `account_type > 1`, the CHECK wasn't meant to ensure that the account_type itself was valid. It was meant to make sure that assuming a valid value, other columns matched it. And FWIW, I'm likely to add an account_type 2 very soon (UIVK only). And the new CHECK that @str4d suggested would have "just worked".